### DISCUSSION TITLE
A placeholder for shared read-only data required by a Directive

### DISCUSSION THREAD

**Discussion by BasicWolf** [sphinx-doc/sphinx#13047] [Q&A] [answered]
Hello, 

I'm looking for a proper placeholder for the data that is collected during a directive's `setup()` call, so that the directive can refer to it during `run()` calls. I.e. something like:

```python
class MyDirective(SphinxDirective):
    def run(self) -> list[nodes.Node]:
        read_my_data()
        ...

def setup(app: Sphinx) -> ExtensionMetadata:
    store_my_data()

    app.add_directive('mydirective', MyDirective)
```

Is there a place in `app` to store data for such purpose?

Thank you!

**Comment by AA-Turner** [comment, marked as answer]
See the [tutorials](https://www.sphinx-doc.org/en/master/development/tutorials/extending_build.html), especially how `todo_all_todos` is managed.

A

**Reply by BasicWolf** [reply]
Thank you, Adam!

Looks like for my case, any event handler which includes `BuildEnvironment` and happens before source-read should do. But I shall rather read carefully about each event intention. 

Thanks again!
